chemprop/v2/data/utils.py

```diff
@@ -1,66 +1,173 @@
+            smiles_indices[smiles] = np.where(all_smiles == smiles)[0]
+
+        # randomly split the unique smiles
+        result = split_fun(np.arange(len(unique_smiles)))
+        train_idxs, val_idxs, test_idxs = _unpack_astartes_result(None, result, include_val)
 
+        # convert these to the 'actual' indices from the original list using the dict we made
+        train = list(itertools.chain.from_iterable(smiles_indices[unique_smiles[i]] for i in train_idxs))
+        val = list(itertools.chain.from_iterable(smiles_indices[unique_smiles[i]] for i in val_idxs))
+        test = list(itertools.chain.from_iterable(smiles_indices[unique_smiles[i]] for i in test_idxs))
+
+    elif split == SplitType.RANDOM:
+        result = split_fun(np.arange(len(data)), sampler="random", **astartes_kwargs)
+        train, val, test = _unpack_astartes_result(data, result, include_val)
+
+    elif split == SplitType.KMEANS:
+        result = mol_split_fun(
+            np.array([m[key_molecule_index] for m in data.smiles()]),
+            sampler="kmeans",
+            hopts=dict(metric="jaccard"),
+            fingerprint="morgan_fingerprint",
+            fprints_hopts=dict(n_bits=2048),
+            **astartes_kwargs,
+        )
+        train, val, test = _unpack_astartes_result(data, result, include_val)
+
+    elif split == SplitType.KENNARD_STONE:
+        result = mol_split_fun(
+            np.array([m[key_molecule_index] for m in data.smiles()]),
+            sampler="kennard_stone",
+            hopts=dict(metric="jaccard"),
+            fingerprint="morgan_fingerprint",
+            fprints_hopts=dict(n_bits=2048),
+            **astartes_kwargs,
+        )
+        train, val, test = _unpack_astartes_result(data, result, include_val)
+
+    else:
+        raise ValueError(f'split type "{split}" not supported.')
+
+    return MoleculeDataset(train), MoleculeDataset(val), MoleculeDataset(test)
+
+
+def _unpack_astartes_result(
+    data: Sequence[MoleculeDatapoint], result: tuple, include_val: bool, log_stats: bool = False
+) -> Tuple[list[MoleculeDatapoint], list[MoleculeDatapoint], list[MoleculeDatapoint]]:
+    """Helper function to partition input data based on output of astartes sampler
+
+    Args:
+        data (MoleculeDataset): The data being partitioned. If None, returns indices.
+        result (tuple): Output from call to astartes containing the split indices
+        include_val (bool): True if a validation set is included, False otherwise.
+        log_stats (bool, optional): Print stats about scaffolds. Defaults to False.
+
+    Returns:
+        MoleculeDatset: The train, validation (can be empty) and test dataset
+    """
+    train_idxs, val_idxs, test_idxs = [], [], []
+    if include_val:
+        train_idxs, val_idxs, test_idxs = result[3], result[4], result[5]
+    else:
+        train_idxs, test_idxs = result[2], result[3]
+    if log_stats:
+        log_scaffold_stats(data, [set(train_idxs), set(val_idxs), set(test_idxs)])
```

Comment:
where is this function defined?

Reply:
Removed this following a comment from Nathan above